File: rr_platform/src/udp_socket/udp_pub.py
#!/usr/bin/env python
import rospy
import socket
from rr_msgs.msg import speed
from rr_msgs.msg import steering
from time import sleep
  
speed_val = 0.0
steering_val = 0.0
buffersize = 1024
port = 8888

# todo make these the right ips
steeringAddressPort = ("192.168.20.3", port)
speedAddressPort = ("192.168.20.4", port)
myAddressPort = ("192.168.20.2", port)

  
def speed_callback(data: speed):
    global speed_val
    speed_val = data.speed
    # print the actual message in its raw format
    rospy.loginfo("Got speed: %s", data.speed)
  
def steering_callback(data: steering):
    global steering_val
    steering_val = data.angle
    # print the actual message in its raw format
    rospy.loginfo("Got steering: %s", data.angle)
  
def main():
    # subscribers
    rospy.init_node('udp_pub', anonymous=True)
    rospy.Subscriber("/speed", speed, speed_callback)
    rospy.Subscriber("/steering", steering, steering_callback)
      
    # create the UDP cient
    socket_client = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    while not rospy.is_shutdown():
        # send the data
        speed_msg = "v={:.2f}".format(speed_val)
        steer_msg = "A={:.2f}".format(steering_val)

        socket_client.sendto(speed_msg.encode(), speedAddressPort)
        socket_client.sendto(steer_msg.encode(), steeringAddressPort)

        rospy.loginfo("Sent speed: %s", speed_msg)
        rospy.loginfo("Sent steering: %s", steer_msg)
        sleep(0.1) # send data at 10Hz

  
if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: rr_platform/src/udp_socket/udp_pub.cpp
#include <ros/ros.h>
#include <rr_msgs/speed.h>
#include <rr_msgs/steering.h>
#include <rr_platform/udp_sock.h>

using std::string;

double cmd_speed = 0;
double cmd_steering = 0;
const string steeringAddress = "192.168.20.3";
const string speedAddress = "192.168.20.4";
const int port = 8888;

void speedCallback(const rr_msgs::speed::ConstPtr& msg) {
    cmd_speed = msg->speed;
}

void steerCallback(const rr_msgs::steering::ConstPtr& msg) {
    cmd_steering = msg->angle;
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "udp_pub_node");
    ros::NodeHandle nh;

    ros::Subscriber speedSub = nh.subscribe("/speed", 1, speedCallback);
    ros::Subscriber steerSub = nh.subscribe("/steering", 1, steerCallback);

    const std::unique_ptr<rr::udp_socket> steeringSocket = std::make_unique<rr::udp_socket>(steeringAddress, port);
    const std::unique_ptr<rr::udp_socket> speedSocket = std::make_unique<rr::udp_socket>(speedAddress, port);

    while (ros::ok()) {
        ros::spinOnce();

        // send speed socket
        string speed_msg = "v=" + std::to_string(cmd_speed);
        speedSocket->send(speed_msg);
        ROS_INFO_STREAM("Sent speed: " << speed_msg);
        // send steering socket
        string steer_msg = "A=" + std::to_string(cmd_steering);
        steeringSocket->send(steer_msg);
        ROS_INFO_STREAM("Sent steering: " << steer_msg);
        ros::Rate rate(10);
    }
}